File: scripts/algorithms/K/Knight Dialer/Knight Dialer.cpp
// Runtime: 48 ms (Top 81.41%) | Memory: 8.50 MB (Top 76.13%)

class Solution {
public:
    static const int mod = 1e9 + 7;
    vector<vector<int>> MOVES = {
    {4, 6},
    {8, 6},
    {7, 9},
    {4, 8},
    {3, 9, 0},
    {},
    {0, 1, 7},
    {2, 6},
    {1, 3},
    {2, 4},
};

    int cache[5001][10];

    int knightDialer(int n) {
        vector<int> nextNumbers = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
        return knightDialer(n, nextNumbers);
    }

    int knightDialer(int remaining, vector<int>& nextNumbers) {
        if (remaining == 1)
            return nextNumbers.size();

        int count = 0;
        for (int nextNumber : nextNumbers) {
            int cur = cache[remaining][nextNumber];
            if (cur == 0) {
                cur = knightDialer(remaining - 1, MOVES[nextNumber]);
                cache[remaining][nextNumber] = cur;
            }
            count += cur;
            count %= mod;
        }
        return count;
    }
};

File: scripts/algorithms/K/Knight Dialer/Knight Dialer.py
// Runtime: 226 ms (Top 89.51%) | Memory: 16.80 MB (Top 56.81%)

class Solution:
    MOD = 10**9 + 7

    def knightDialer(self, n: int) -> int:
        # Initialize an array to store the possible positions of the knight on the phone pad
        cur_pos = [1] * 10

        # Loop through the number of jumps required
        for jump in range(2, n + 1):
            # Create a new list to store the updated positions after each jump
            new_pos = [0] * 10

            # Calculate the new positions based on the valid knight moves
            new_pos[0] = (cur_pos[6] + cur_pos[4]) % self.MOD
            new_pos[1] = (cur_pos[6] + cur_pos[8]) % self.MOD
            new_pos[2] = (cur_pos[7] + cur_pos[9]) % self.MOD
            new_pos[3] = (cur_pos[4] + cur_pos[8]) % self.MOD
            new_pos[4] = (cur_pos[0] + cur_pos[3] + cur_pos[9]) % self.MOD
            new_pos[5] = 0  # Knight cannot move to position 5
            new_pos[6] = (cur_pos[0] + cur_pos[1] + cur_pos[7]) % self.MOD
            new_pos[7] = (cur_pos[2] + cur_pos[6]) % self.MOD
            new_pos[8] = (cur_pos[1] + cur_pos[3]) % self.MOD
            new_pos[9] = (cur_pos[2] + cur_pos[4]) % self.MOD

            # Update the current positions list for the next iteration
            cur_pos = new_pos

        # Calculate the total count of distinct phone numbers
        total_count = sum(cur_pos) % self.MOD

        return total_count
